programs/marginfi/src/ix_utils.rs

```diff
@@ -0,0 +1,262 @@
+}
+
+/// The function of struct discrminator is constructed from these 8 bytes. Typically, the namespace
+/// is "account" or "state"
+///
+/// e.g. for LiquidateStart:
+/// ```
+///  let discrim = get_function_hash("global", "liquidate_start")
+/// ```
+pub fn get_discrim_hash(namespace: &str, name: &str) -> [u8; 8] {
+    let preimage = format!("{}:{}", namespace, name);
+    let mut sighash = [0u8; 8];
+    sighash.copy_from_slice(
+        &anchor_lang::solana_program::hash::hash(preimage.as_bytes()).to_bytes()[..8],
+    );
+    sighash
+}
+
+/// Validate the given ix hash is the first in the list of ixes, other than compute budget ixes,
+/// and appears only once.
+///
+/// If the ix implements `Hashable`, use `get_hash()` to get the expected hash.
+pub fn validate_ix_first(
+    ixes: &[Instruction],
+    program_id: &Pubkey,
+    expected_hash: &[u8],
+) -> MarginfiResult<()> {
+    let compute_budget_key = COMPUTE_PROGRAM_KEY;
+    let mut first_non_compute_ix_encountered = false;
+
+    for instruction in ixes.iter() {
+        if instruction.program_id == compute_budget_key {
+            continue;
+        }
+
+        // The remainder of the data is the args, which don't matter for validation
+        let ix_data_func = &instruction.data[0..8];
+
+        // If this is the first non-compute ix, it is either the setup ix, or we fail
+        if !first_non_compute_ix_encountered {
+            if instruction.program_id == *program_id && ix_data_func == expected_hash {
+                first_non_compute_ix_encountered = true;
+            } else {
+                msg!(
+                    "Expected ix from program: {:?} w/ hash: {:?}",
+                    program_id,
+                    expected_hash
+                );
+                msg!(
+                    "Got ix from program: {:?} w/ hash: {:?}",
+                    instruction.program_id,
+                    instruction.data,
+                );
+                msg!("Start IX was not the first ix (other than compute).");
+                return err!(MarginfiError::StartNotFirst);
+            }
+        } else {
+            // Already validated the first ix, check that the start ix does not appear again
+            if instruction.program_id == *program_id && ix_data_func == expected_hash {
+                msg!("Setup IX appears more than once.");
+                return err!(MarginfiError::StartRepeats);
+            }
+        }
+    }
+
+    if first_non_compute_ix_encountered {
+        Ok(())
+    } else {
+        msg!("Start IX was not found in the TX.");
+        err!(MarginfiError::StartNotFirst)
+    }
+}
+
+/// Validate the given ix hash is the last ix in the list of ixes.
+pub fn validate_ix_last(
+    ixes: &[Instruction],
+    program_id: &Pubkey,
+    expected_hash: &[u8],
+) -> MarginfiResult<()> {
+    let last_ix = ixes.last().unwrap(); // Safe unwrap, ixes.size() >= 1
+    if !last_ix.program_id.eq(program_id) || last_ix.data != expected_hash {
```

Comment:
Maybe, we're thinking of changing how we discriminators, but should probably be explicitly checking the first eight bytes of the `data[..8]` and the expected hash should be also an 8 byte array.